pkg/services/ngalert/state/cache.go

```diff
@@ -60,54 +60,33 @@ func (rs *ruleStates) getOrCreate(ctx context.Context, log log.Logger, alertRule
 	labels, _ := expand(ctx, log, alertRule.Title, alertRule.Labels, templateData, externalURL, result.EvaluatedAt)
 	annotations, _ := expand(ctx, log, alertRule.Title, alertRule.Annotations, templateData, externalURL, result.EvaluatedAt)
 
-	values := make(map[string]float64)
-	for refID, v := range result.Values {
-		if v.Value != nil {
-			values[refID] = *v.Value
-		} else {
-			values[refID] = math.NaN()
-		}
+	labels, duplicateLabels := mergeLabelsKeepDuplicates(extraLabels, labels)
+	if len(duplicateLabels) > 0 {
+		log.Warn("Custom labels conflict with one or more reserved labels, the duplicate labels will be ignored",
+			"reserved_labels", extraLabels, "duplicate_labels", duplicateLabels)
 	}
 
-	lbs := make(data.Labels, len(extraLabels)+len(labels)+len(result.Instance))
-	dupes := make(data.Labels)
-	for key, val := range extraLabels {
-		lbs[key] = val
-	}
-	for key, val := range labels {
-		ruleVal, ok := lbs[key]
-		// if duplicate labels exist, reserved label will take precedence
-		if ok {
-			if ruleVal != val {
-				dupes[key] = val
-			}
-		} else {
-			lbs[key] = val
-		}
-	}
-	if len(dupes) > 0 {
-		log.Warn("Rule declares one or many reserved labels. Those rules labels will be ignored", "labels", dupes)
-	}
-	dupes = make(data.Labels)
-	for key, val := range result.Instance {
-		_, ok := lbs[key]
-		// if duplicate labels exist, reserved or alert rule label will take precedence
-		if ok {
-			dupes[key] = val
-		} else {
-			lbs[key] = val
-		}
-	}
-	if len(dupes) > 0 {
-		log.Warn("Evaluation result contains either reserved labels or labels declared in the rules. Those labels from the result will be ignored", "labels", dupes)
+	labels, duplicateLabels = mergeLabelsKeepDuplicates(labels, result.Instance)
+	if len(duplicateLabels) > 0 {
+		log.Warn("Evaluation labels conflict with one or more custom or reserved labels, the duplicate labels will be ignored",
+			"labels", labels, "duplicate_labels", duplicateLabels)
```

Comment:
The only problem I have with this solution is that it copies too many times. Each call of `mergeLabelsKeepDuplicates` creates a new map and copies labels to it and then returns map. At the end, we will have 1 more map than the current code does. I think we should just create a function that merges 3 maps into one and returns duplicates (2 sets if we want to distinguish what map had dupes).

Reply:
We don't do a huge number of copies though. We create at most two maps and then copy O(2N) slice headers into already allocated memory. We don't copy the full string, just the 16 bit headers.

Reply:
We don't do a huge number of copies though. We create two new map[string]string and then "move" O(2N) slice headers into pre-allocated memory. We don't copy the every string, just the 24 bit slice headers?

Reply:
We don't do a huge number of copies though. We create two new map[string]string and then "move" O(2N) slice headers into the already allocated map. We don't copy the every string, just each 24 bit string header?

Reply:
The current implementation does the following:

- create map `result` of size `a`+`b`+`c`
- copy map `a` to `result`
- create `dupes` of 0 size
- iterate `b` and copy only non-dupes to `result`, save dupes to `dupes`
- reset `dupes`
- iterate `c` and copy only non-dupes to `result`, save dupes to `dupes`

The proposed change does

- create map `result` of size `a`+`b`
- copy `a` to `result`
- create `dupes` of 0 size
- iterate `b` and copy only non-dupes to `result`, save dupes to `dupes`
- create `new result` of size `result` + `c`
- copy `result` to `new result` 
- iterate `c` and copy only non-dupes to `result`, save dupes to `dupes`

So, it adds 1 extra copy operation of the map. It's fast and considering that label maps aren't usually big, the impact would not be noticeable. I just do not see any benefits of the proposed solutions so it can be a reason for the change. 

The function is not reused anywhere else. Do you have plans on reusing this function? 

I think just extracting the logic as is to simplify `getOrCreate` could be an alternative. Or a function that mutates the first map and just skips dupes (it's not pure function but I think that's fine) it can even be an anonymous one. 

It's just my 2c. I do not have a strong opinion against this refactor.